File: 加权轮询.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

struct Server {
  string name;
  int weight;
};

class LoadBanlancer {
private:
  vector<string> virtualNodes; // 虚拟节点列表
  vector<Server> servers;
  int current;

public:
  LoadBanlancer(vector<Server> server) : current(-1) {
    this->servers = server;

    for (const auto &server : servers) {
      for (int i = 0; i < server.weight; i++) {
        virtualNodes.push_back(server.name);
      }
    }
    random_shuffle(virtualNodes.begin(), virtualNodes.end());
  }

  string getServer() {
    if (virtualNodes.empty())
      return "";
    current = (current + 1) % virtualNodes.size();
    return virtualNodes[current];
  }
};

int main() {

  vector<Server> server = {{"A", 5}, {"B", 3}, {"C", 2}};

  LoadBanlancer lb(server);

  //测试20次

  for (int i = 0; i < 20; i++) {
    cout << "Request " << i + 1 << " goes to: " << lb.getServer() << endl;
  }
}

File: 加权轮询.go
package main


import "fmt"

type Server struct{
	name string
	weight int
}

type LoadBanlancer struct {
	virtualNodes []string
	current int
	servers map[string]int
}

func NewLoadBalancer(servers map[string]int) *LoadBanlancer{
	servers=servers


	var nodes []string
	for serverName,wight :=range servers{
		for i:=0;i<wight;i++{
			nodes=append(nodes, serverName)
		}
	}

	return &LoadBanlancer{
		virtualNodes: nodes,
		current: -1,
	}
}

func (lb *LoadBanlancer)GetServer() string{
	if len(lb.virtualNodes)==0{
		return ""
	}

	lb.current=(lb.current+1)%len(lb.virtualNodes)
	return lb.virtualNodes[lb.current]
}


func main(){

	servers := map[string]int{
		"A": 5,
		"B": 3,
		"C": 2,
	}

	lb:=NewLoadBalancer(servers)

	for i:=0;i<20;i++{
		fmt.Printf("Request %d goes to: %s\n",i+1,lb.GetServer())
	}
}